Move C global variables to a dedicated structure

kayceesrk opened the pull request:
This PR moves C global variables in the runtime to a dedicated structure, a "domain state" table. This PR is a pre-requisite for the multicore runtime, where each domain (a parallel thread of execution) will need to maintain its own table of domain local variables. This PR does not introduce any API changes. 

In order to allow fast access to the domain state, we steal the exception pointer register (r14 on amd64) and make it point to the domain state table. The exception pointer is now a field in the domain state table. Convenience macros are introduced to access the domain state fields in the runtime and in the backend of the compiler. The PR is syntactically fairly invasive. But the semantics changes are fewer.

## Status

~~Currently, only exception pointer, young_pointer, and young_limits are in the domain state. The only tested and working architecture is linux on amd64. The build is expected to fail on every other configuration.~~

- [x] Introduce domain state and steal exception pointer register
- [x] Move `young_ptr` and `young_limit` to domain state
- [x] Move multicore relevant native runtime C globals to the domain state
- [x] Move multicore relevant bytecode runtime C global to the domain state
- [x] Support Linux x86-64
- [x] Support macOS x86-64
- [x] Support Windows cygwin x86/x86-64
- [x] Support Windows mingw x86-64
- [x] Support FreeBSD x86-64
- [x] Support Linux x86
- [x] Support Windows x86
- [x] Support Linux arm64
- [x] Support Linux arm32
- [x] Support Linux Power64
- [x] Support Linux IBM Z
- [x] Support Windows min x86-64
- [x] Support Windows Native (Microsoft Visual C/C++) x86-64

## Changes

### Exceptions

On amd64, exception pointers are cached in the r14 register. In this code snippet:

```ocaml
exception Foo

let foo () =
  try ()
  with Foo -> ()

let bar () = raise Foo
```

the diff in the push trap and pop trap compilation is:

```diff
 .L103:
-       .cfi_adjust_cfa_offset 16
-       subq    $16, %rsp
-       movq    %r14, (%rsp)
-       leaq    .L102(%rip), %r14
-       movq    %r14, 8(%rsp)
-       movq    %rsp, %r14
+       leaq    .L102(%rip), %r11
+       .cfi_adjust_cfa_offset 8
+       pushq   %r11
+       .cfi_adjust_cfa_offset 8
+       pushq   8(%r14)
+       movq    %rsp, 8(%r14)
        movq    $1, %rax
-       popq    %r14
+       popq    8(%r14)
        .cfi_adjust_cfa_offset -8
        addq    $8, %rsp
        .cfi_adjust_cfa_offset -8
```

The exception pointer is stored at `8(%r14)`. The diff in the raise code is as expected:

```diff
 .L104:                                                    
        movq    camlHello@GOTPCREL(%rip), %rax             
        movq    (%rax), %rax                               
-       movq    %r14, %rsp                                 
-       popq    %r14                                       
+       movq    8(%r14), %rsp                              
+       popq    8(%r14)                                    
        popq    %r11                                       
        jmp     *%r11                                      
        .cfi_endproc
```

### Allocations

Inlined allocations sequences are shorter since the `young_limit` is now in the domain state. The diff in the allocation sequence in the code:

```ocaml
let baz x = (x,x)
```

is

```diff
        movq    %rax, %rbx
 .L101:
        subq    $24, %r15
-       movq    caml_young_limit@GOTPCREL(%rip), %rax
-       cmpq    (%rax), %r15
+       cmpq    (%r14), %r15
        jb      .L102
        leaq    8(%r15), %rax
        movq    $2048, -8(%rax)
```

`young_limit` is stored at `0(%r14)`.

### FFI

There is no need to cache and flush the exception pointer when calling C functions. The fact that the exception pointer is not available as a global variable means that functions in `amd64.S` such as `caml_start_program`, callbacks, functions for raising exceptions will need the domain state as an extra argument. This change is internal only and the C API remains the same. 

## Performance 

The goal is to retain the performance with respect to trunk. Preliminary performance numbers are available for [micro](http://bench.ocamllabs.io/comparison/?exe=14%2BL%2Br14-globals&ben=1%2C2%2C110%2C111%2C5%2C6%2C7%2C9%2C8%2C10%2C12%2C11%2C15%2C16%2C17%2C13%2C14%2C20%2C21%2C22%2C18%2C19%2C25%2C26%2C27%2C23%2C24%2C30%2C31%2C32%2C28%2C29%2C33%2C152%2C151%2C37%2C112%2C113%2C114%2C115%2C116%2C117%2C118%2C119%2C120%2C122%2C121%2C123%2C124%2C125%2C126%2C127%2C54%2C154%2C155%2C153%2C157%2C158%2C156%2C161%2C162%2C163%2C164%2C159%2C160%2C167%2C168%2C169%2C170%2C165%2C166%2C172%2C171%2C174%2C173%2C86&env=2&hor=true&bas=12%2BL%2Btrunk&chart=normal+bars) and [macro](http://bench2.ocamllabs.io/comparison/?exe=10%2BL%2Br14-globals&ben=1%2C2%2C130%2C131%2C3%2C4%2C5%2C6%2C7%2C8%2C9%2C10%2C11%2C12%2C13%2C75%2C76%2C77%2C78%2C79%2C80%2C81%2C82%2C83%2C84%2C132%2C133%2C14%2C15%2C16%2C85%2C86%2C87%2C88%2C17%2C18%2C19%2C20%2C21%2C22%2C23%2C24%2C25%2C26%2C89%2C90%2C91%2C92%2C93%2C94%2C27%2C28%2C29%2C30%2C31%2C32%2C33%2C34%2C35%2C36%2C37%2C38%2C39%2C40%2C41%2C42%2C95%2C43%2C44%2C45%2C46%2C47%2C48%2C49%2C50%2C51%2C52%2C53%2C54%2C55%2C96%2C97%2C98%2C99%2C56%2C100%2C57%2C58%2C59%2C60%2C61%2C101%2C62%2C63%2C64%2C65%2C66%2C102%2C67%2C68%2C69%2C70%2C129%2C103%2C104%2C105%2C106%2C107%2C108%2C71%2C72%2C74%2C109%2C110%2C111%2C112%2C113%2C114%2C115%2C116%2C117%2C118%2C119%2C120%2C121%2C122%2C123%2C124%2C125%2C126%2C127%2C128%2C73&env=3&hor=true&bas=9%2BL%2Btrunk&chart=normal+bars) benchmarks. There are a few outliers (in either direction) which I will investigate. And a few, which look like spurious runs (`sequence_cps` in the macrobenchmark). 

## Reviewing

It is instructive to start the review at `domain_state.tbl` which introduces the domain state table. The OCaml and C macros for accessing domain state are generated in `utils/domainstate.ml*` and `runtime/caml/domain_state.h`. The interesting changes are in the files `asmcomp/amd64/emit.mlp` and `runtime/amd64.S`.

kayceesrk committed:
Introduce domain state and steal exception pointer

kayceesrk committed:
young_ptr and young_limit are now in domain state

kayceesrk committed:
Destroy r11 at Itrywith

kayceesrk committed:
Reorder domain state fields

kayceesrk committed:
Make depend and fix typos

alainfrisch commented:
- With this change, how far would be we from being able to run multiple independant OCaml runtimes in different threads of the same process?  (Exposed e.g. as a way to "restart" the main OCaml entry point in a different thread.)

- In addition to globals that belong to the runtime system itself, C bindings can also use globals, e.g. to cache "named OCaml values/exceptions".  Do you plan to expose an API to allow external C code to store custom data in the "domain state"?

c-cube commented:
@alainfrisch that's a very interesting idea; would it even be possible to push it further and have both native and bytecode runtimes in the same process, on different domains?

kayceesrk commented:
> With this change, how far would be we from being able to run multiple independant OCaml runtimes in different threads of the same process? (Exposed e.g. as a way to "restart" the main OCaml entry point in a different thread.)

IIUC in this runtime, each thread's heap (minor and major) is completely independent of the other threads? Such a design is not part of the multicore OCaml plan. But I believe it would take a little more work to have such a feature. Specifically, handling signals (pthreads share signal dispositions), fork (threads and fork don't mix well), global roots (multicore has a single global root list, this runtime will have a global root list per domain), etc. Some of these issues will need to be handled in the multicore runtime as well.

> In addition to globals that belong to the runtime system itself, C bindings can also use globals, e.g. to cache "named OCaml values/exceptions". Do you plan to expose an API to allow external C code to store custom data in the "domain state"?

Multicore support backwards compatible `caml_name_value` API (See https://github.com/ocaml-multicore/ocaml-multicore/commit/8d7bc72dd02811d72fe7118629007b11bcb5180a). There are no plans to extend this API.

kayceesrk commented:
@c-cube I haven't thought through this problem enough. But one problem I see in this approach is build tools; it is unclear now whether bytecode or native versions of libraries must be linked. Do you have a use-case in mind for mixing bytecode and native runtimes?

c-cube commented:
My use case is quite niche, but it's about having both a bytecode program that embeds OCaml's interpreter (and extends it) on the one hand (so it must be bytecode), and efficient native code on the other hand. Currently the closest I can do is multi-process and message passing, and domains could make it a bit simpler. I don't think it's particularly useful outside of use cases mixing heavy computations and embedding a toplevel.

bluddy commented:
One example is an editor written in ocaml that also uses interpreted ocaml for configuration etc.

kayceesrk commented:
Thanks @c-cube and @bluddy. Those are indeed interesting use cases. I'll keep this in the back of my mind while hacking on the runtime system changes for multicore. But I am afraid this feature is not something that I have the time to look at right now.

kayceesrk commented:
@ctk21 kindly analyzed the performance difference in `sequence_cps` and the difference we see is consistent with the variance that we had observed earlier due to code layout changes. The [timeline](http://bench2.ocamllabs.io/timeline/?exe=2%2C5%2C8%2C9%2C10&base=&ben=sequence-cps&env=3&revs=50&equid=off&quarts=on&extr=on) shows that the current performance difference between trunk and this feature branch is within the historical noise.

alainfrisch commented:
> IIUC in this runtime, each thread's heap (minor and major) is completely independent of the other threads?

Yes, absolutely.

> Such a design is not part of the multicore OCaml plan. But I believe it would take a little more work to have such a feature. Specifically, handling signals (pthreads share signal dispositions), fork (threads and fork don't mix well), global roots (multicore has a single global root list, this runtime will have a global root list per domain), etc. Some of these issues will need to be handled in the multicore runtime as well.

Wouldn't global roots be stored in the domain state as well, as part of this PR?

kayceesrk commented:
> Wouldn't global roots be stored in the domain state as well, as part of this PR?

Not the global roots. The intention is to only move the domain local C global variables to the domain state. Since the global roots are shared between the domains, the global roots structure remains a C global variable in multicore: https://github.com/ocaml-multicore/ocaml-multicore/blob/master/byterun/globroots.c#L47-L48. I intend to leave the global roots as a global variable. 

However, if we wanted a thread-local runtime at some point in the future, then we can move the global roots as a field in the domain state, with the field pointing to the same global roots list in the case of shared heap multicore and different roots list for thread-local rutime. This should be straight forward.

kayceesrk committed:
macOS support

chambart commented:
One use case I would have for such a completely separated runtime would be to allow writing clean C interfaces for OCaml libraries without hidden global variables. Of course that would require also some changes to the API.

kayceesrk committed:
Various minor heap pointers moved to domain state

jordwalke commented:
I understand this PR alone doesn't fully implement support for what @c-cube asked about (running multiple OCaml runtimes, each having their own thread in one process). There are many great use cases for this. For example, the Oni editor written with native OCaml/Reason that has a high priority main thread to do UI drawing, but would like to support extensions/plugins which would be extended and mostly isolated from the core editor UI thread.

The other use cases I have is GUI applications that have one thread doing the interactions and animations, and another thread that handles lower priority data processing that shouldn't block the app. They often don't need to share much memory - and it is suitable to create C bindings that lock some serializable resource for communication. Currently we can use OCaml threads, but then there are long periods of time that cores go unutilized.

kayceesrk committed:
Move all statistics variables to domain state

kayceesrk committed:
Move backtrace support global variables to domain state.

Since we cannot access backtrace position in cmmgen.ml anymore,
Cmm.raise_kind in removed. Instead, we use Lambda.raise_kind. When
assembly code is generated, we reset the backtrace position to 0 in the
case of regular raise. Importantly, the semantics remains the same.

kayceesrk committed:
Move bytecode global variables to domain state

kayceesrk committed:
Move native runtime globals used for roots to domain state

kayceesrk committed:
Move local_roots, compare unorderd and gc request variables to domain state

kayceesrk commented:
All of the domain local C global variables have been moved to the domain state. For other C global variables, I have documented below why they have not been moved. This might be useful for someone implementing the thread-local runtime proposed here. The next step is to implement the change in the different backends. 

### C globals not moved to domain state

Hooks are installed at the process level. For multicore, we will change the signatures of these functions to take `Caml_state` as an extra parameter. Using this, the hooks can identify which domains it was invoked for.

* caml_natdynlink_hook
* caml_termination_hook
* caml_scan_roots_hook
* caml_stack_usage_hook
* caml_finalise_begin_hook
* caml_finalise_end_hook
* caml_major_slice_begin_hook
* caml_major_slice_end_hook
* caml_minor_gc_begin_hook
* caml_minor_gc_end_hook
* caml_major_gc_hook
* caml_enter_blocking_section_hook
* caml_leave_blocking_section_hook
* caml_async_action_hook
* caml_sigmask_hook
* caml_try_leave_blocking_section_hook

Global roots are shared between the domains.

* caml_global_roots
* caml_global_roots_old
* caml_global_roots_young
* caml_globals_inited
* caml_globals_scanned
* caml_dyn_globals
* caml_incremental_roots_count

Channel list + hooks are for the entire process.

* caml_all_opened_channels
* caml_channel_mutex_free
* caml_channel_mutex_lock
* caml_channel_mutex_unlock
* caml_channel_mutex_unlock_exn
* channel_operations

AFL configuration variables are for the entire process. The details of afl operations under multicore needs to be worked out.

* caml_abort_on_uncaught_exn
* afl_area_initial
* afl_initialised
* caml_afl_area_ptr
* caml_afl_prev_loc

Free list. Multicore uses a diffent allocator. Hence, these variables will not be moved.

* beyond
* flp
* flp_size
* caml_fl_wsz_at_phase_change
* caml_fl_cur_wsz
* caml_fl_merge
* fl_prev
* fl_last
* sentinel

Major GC is different in multicore.

* caml_gc_phase
* caml_gc_subphase
* caml_gc_sweep_hp
* caml_gc_clock
* caml_heap_start
* caml_major_ring
* caml_major_ring_index
* caml_major_window
* caml_major_work_credit
* gray_vals
* gray_vals_cur
* gray_vals_end
* gray_vals_size
* heap_is_pure
* caml_allocated_words
* caml_dependent_size
* caml_dependent_allocated
* caml_extra_heap_resources
* markhp
* chunk
* limit
* p_backlog
* caml_allocation_policy

Runtime parameters apply to all the domains:

* caml_parser_trace
* caml_percent_free
* caml_percent_max
* caml_major_heap_increment
* caml_init_custom_major_ratio
* caml_init_custom_minor_max_bsz
* caml_init_custom_minor_ratio
* caml_init_heap_chunk_sz
* caml_init_heap_wsz
* caml_init_major_window
* caml_init_max_percent_free
* caml_init_max_stack_wsz
* caml_init_minor_heap_wsz
* caml_init_percent_free
* caml_cleanup_on_exit
* caml_trace_level
* caml_verb_gc
* caml_use_huge_pages
* caml_custom_major_ratio
* caml_custom_minor_max_bsz
* caml_custom_minor_ratio
* caml_runtime_warnings
* caml_runtime_warnings_first

The details of interaction of profilers and multi-threaded execution has not been worked out yet. If the profiles can be collected per domain and collated later, then we can move these variables under domain state.

* caml_memprof_postponed_head
* caml_memprof_suspended
* caml_memprof_young_trigger
* caml_spacetime_trie_node_ptr
* caml_spacetime_finaliser_trie_root
* caml_extern_allow_out_of_heap

I can't see this variable used in the runtime.

* caml_huge_fallback_count

Process level constants.

* caml_code_area_end
* caml_code_area_start
* caml_exe_name
* caml_system__frametable
* frametables
* caml_frame_descriptors
* caml_frame_descriptors_mask
* caml_int32_ops
* caml_nativeint_ops
* caml_int64_ops
* caml_ba_ops
* caml_code_fragments_table
* caml_atom_table
* caml_array_bound_error_exn
* caml_ba_element_size
* caml_debug_info

Signals are handled at process-level. Any domain may be interrupted to deliver the signal.

* caml_async_signal_mode
* caml_pending_signals
* caml_signal_handlers
* caml_signals_are_pending

No page tables in multicore.

* caml_page_table

The debugger will only work for single-threaded execution due to fork.

* caml_debugger_fork_mode
* caml_debugger_in_use
* caml_event_count

Multicore uses a different ephemeron implementation and the structures are different. Not moving existing structures now.

* caml_ephe_list_head
* caml_ephe_none
* ephe_dummy
* ephe_list_pure
* ephes_checked_if_pure
* ephes_to_check

--

The following items were added after @gasche [pointed out](https://github.com/ocaml/ocaml/pull/8713#issuecomment-526353663) that this list is not exhaustive. 

Marshaling has (various problems in multicore)[https://github.com/ocaml-multicore/ocaml-multicore/issues/75]. Since the design isn't final yet, I have not yet moved the following variables to the domain state. 

* extern_flags
* extern_limit
* extern_output_block
* extern_output_first
* extern_ptr
* extern_stack_init
* extern_trail_block
* extern_trail_cur
* extern_trail_first
* extern_trail_limit
* extern_userprovided_output
* intern_block
* intern_color
* intern_dest
* intern_extra_block
* intern_header
* intern_input
* intern_obj_table
* intern_src
* intern_stack_init

kayceesrk committed:
Rename exn_handler to exception_pointer

kayceesrk committed:
Support for domain state on arm64 architecture

kayceesrk committed:
Restore exception pointer macros in signals_osdeps.h

kayceesrk committed:
Domain state works on Power64

kayceesrk commented:
The domain state now works on Linux+arm64 and Linux+power. This PR should also work on power (32bit), but is untested. On arm64 and power, there are enough registers available. Hence, we use the callee saved register r28 on power and x25 on arm64 for the domain state pointer instead of taking over exception pointer register (r14 in amd64). Obviously, r28 on power and x25 on arm64 are not available for register allocation.

kayceesrk committed:
Add domain_state.tbl to install targets

kayceesrk committed:
Fix domain state field offset for power (32-bit)

The domain state fields are always aligned at 8 byte offset. This is to
ensure that even on a 32-bit where pointers are 32-bits and doubles are
64-bits, the offset calculation remains the same as 64-bit
architectures.

kayceesrk committed:
Domain state support for i386

kayceesrk committed:
Optimising poptrap in i386

kayceesrk commented:
Implemented support for `i386`. Tested Windows (Cygwin) + x86-64. Cygwin x86 should work since there are not windows specific things in the i386 port. Also tested on FreeBSD x86-64.

kayceesrk committed:
Support for domain state on arm

kayceesrk commented:
`arm` is now supported.

kayceesrk committed:
Domain state support for s390x

kayceesrk commented:
`s390x` now works.

kayceesrk committed:
Fix long lines

kayceesrk committed:
Add changes entry

kayceesrk committed:
Fix references to removed variables

kayceesrk committed:
Intialise domains earlier

kayceesrk committed:
Define alignment macro for MSVC compiler

kayceesrk committed:
Fix CAML_STATIC_ASSERT for MSVC

dra27 commented:
I see the AppVeyor failure with Visual Studio 2015 but not with Visual Studio 2017. This patch appears to resolve it for VS2015:

```diff
diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index 9164d1098..e11b1673f 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -125,7 +125,7 @@ extern caml_timing_hook caml_minor_gc_begin_hook, caml_minor_gc_end_hook;
 extern caml_timing_hook caml_finalise_begin_hook, caml_finalise_end_hook;

 #define CAML_STATIC_ASSERT_3(b, l) \
-  typedef CAMLunused_start \
+  CAMLunused_start typedef \
     char static_assertion_failure_line_##l[(b) ? 1 : -1] CAMLunused_end
 #define CAML_STATIC_ASSERT_2(b, l) CAML_STATIC_ASSERT_3(b, l)
 #define CAML_STATIC_ASSERT(b) CAML_STATIC_ASSERT_2(b, __LINE__)
```

however both VS2015 and VS2017 eventually hit:

```
OCAML_FLEXLINK="./boot/ocamlrun ./flexdll/flexlink.exe" ./boot/ocamlrun ./ocamlopt -g -nostdlib -I stdlib -I otherlibs/dynlink  -o ocamlc.opt compilerlibs/ocamlcommon.cmxa compilerlibs/ocamlbytecomp.cmxa driver/main.cmx -cclib "advapi32.lib ws2_32.lib version.lib"
** Cannot resolve symbols for stdlib\libasmrun.lib(callback_n.obj):
 caml_callback2_asm
 caml_callback3_asm
 caml_callback_asm
** Cannot resolve symbols for stdlib\libasmrun.lib(amd64nt.obj):
 caml_backtrace_active
 caml_backtrace_pos
 caml_bottom_of_stack
 caml_exception_pointer
 caml_gc_regs
 caml_last_return_address
 caml_young_limit
 caml_young_ptr
File "caml_startup", line 1:
Error: Error during linking
make[2]: *** [Makefile:885: ocamlc.opt] Error 2
make[2]: Leaving directory '/cygdrive/c/Users/DRA/Documents/ocaml-multicore'
make[1]: *** [Makefile:424: opt.opt] Error 2
make[1]: Leaving directory '/cygdrive/c/Users/DRA/Documents/ocaml-multicore'
make: *** [Makefile:475: world.opt] Error 2
```
The changes to `amd64.S` need applying to `amd64nt.asm`

kayceesrk committed:
Fix CFI offsets in amd64

kayceesrk committed:
Domain state support for msvc64

kayceesrk committed:
Add copyright header to gen_domain_state_inc.awk

kayceesrk commented:
msvc64 now works. The CI tests are now passing.

kayceesrk commented:
@alainfrisch @chambart Would be interested in hearing our thoughts on the PR.

dra27 commented on Makefile:
```diff
@@ -893,6 +898,10 @@ ocamlopt.opt: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
               $(OPTSTART:.cmo=.cmx)
 	$(CAMLOPT_CMD) $(LINKFLAGS) -o $@ $^
 
+ocamlopt.d.opt: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
```
Does this new target need to remain (I think it needs cleaning, if so)

kayceesrk replied:
Done.

nojb commented:
precheck build: https://ci.inria.fr/ocaml/job/precheck/273/

dra27 commented on runtime/caml/domain.h:
```diff
@@ -0,0 +1,37 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
```
This isn't right!

kayceesrk replied:
Fixed.

dra27 commented on runtime/caml/domain_state.h:
```diff
@@ -0,0 +1,47 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
```
Similarly!

kayceesrk replied:
Fixed.

dra27 commented on runtime/caml/domain.h:
```diff
@@ -0,0 +1,37 @@
+
+#endif /* CAML_INTERNALS */
+
+CAMLextern caml_domain_state* Caml_state;
```
This I think will result in an error if `CAML_INTERNALS` isn't defined - should it be within the block too?

kayceesrk replied:
Removed this extern declaration as it was superfluous; `domain_state.h` declares the extern which is included in this file.

dra27 commented on runtime/caml/domain_state.tbl:
```diff
@@ -0,0 +1,72 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
```
Another header!

kayceesrk replied:
Fixed.

dra27 commented on runtime/domain.c:
```diff
@@ -0,0 +1,82 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
```
Another header to do

kayceesrk replied:
Fixed.

dra27 commented on runtime/gen_domain_state_inc.awk:
```diff
@@ -0,0 +1,34 @@
+#*                                                                        *
+#*                                 OCaml                                  *
+#*                                                                        *
+#*         Damien Doligez, projet Gallium, INRIA Rocquencourt             *
```
Header to do

kayceesrk replied:
Fixed.

dra27 commented on utils/domainstate.ml.c:
```diff
@@ -0,0 +1,36 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
```
Another header

kayceesrk replied:
Fixed.

dra27 commented on utils/domainstate.mli.c:
```diff
@@ -0,0 +1,6 @@
+type t =
```
This file should probably have a header too

kayceesrk replied:
Done.

dra27 commented on Makefile:
```diff
@@ -333,12 +331,19 @@ reconfigure:
 	./configure $(CONFIGURE_ARGS)
 endif
 
+utils/domainstate.ml: utils/domainstate.ml.c runtime/caml/domain_state.tbl
```
These utils/domainstate.ml and utils/domainstate.mli need to go in `.gitignore`

kayceesrk replied:
Done.

dra27 commented on runtime/Makefile:
```diff
@@ -353,7 +353,10 @@ $(foreach object_type, $(object_types), \
 	$(ASPP) $(ASPPFLAGS) $(SHAREDLIB_CFLAGS) -o $@ $<
 
-%.obj: %.asm
+domain_state.inc: caml/domain_state.tbl gen_domain_state_inc.awk
```
`runtime/domain_state.inc` needs to go in `.gitignore`

kayceesrk replied:
Done.

dra27 commented on runtime/caml/compatibility.h:
```diff
@@ -235,10 +235,6 @@
 /* **** meta.c */
 
 /* **** minor_gc.c */
```
Is this huge change the appropriate time to consider removing `compatibility.h` (or at least making its inclusion a preprocessor error)?

damiendoligez replied:
It may be the right time, but this should be a separate PR, that we'll need to check on OPAM and be ready to revert.

kayceesrk replied:
ugh. The macro definitions such as `#define young_ptr ....` interferes with the field with the same name in domain state. If we are not modifying `compatibitlity.h` in this PR, then we should rename the domain state fields (`young_start`, `young_end`, `young_ptr`, `young_limit`, `local_roots`) to something else (by perhaps adding a prefix). Any suggestions on this?

kayceesrk replied:
I reverted the change to `compatibility.h`. Now the PR doesn't modify compatibility.h.

dra27 reviewed (commented):
Build system and copyright headers - I haven't reviewed the technical details.

kayceesrk committed:
Domain state support for msvc32

kayceesrk committed:
Add missing awk file

kayceesrk committed:
Fix makefile, gitignore and a stray extern declaration

kayceesrk committed:
Fix tests/asmgen for msvc32

kayceesrk committed:
Copyright headers and attribution

kayceesrk committed:
Fix copytight header in awk files

kayceesrk commented:
Precheck passes: https://ci.inria.fr/ocaml/job/precheck/275/

The one failing test on mingw32 is the random number generation test which is expected to fail every 10000 runs [0].  

[0] https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/lib-random/rand.ml

kayceesrk committed:
Fix copyright headers again

kayceesrk commented:
All prechecks are successful https://ci.inria.fr/ocaml/job/precheck/276/

dra27 commented on Changes:
```diff
@@ -37,14 +37,17 @@ Working version
    review by Xavier Leroy and Gabriel Scherer)
 
-* #8711: The major GC hooks are no longer allowed to interact with the
+- #8711: The major GC hooks are no longer allowed to interact with the
```
Is this meant to be here?

kayceesrk replied:
Yes. Makes the Changes file consistent. Other items use `-` while this one uses `*`. Happy to remove it.

dra27 replied:
The asterisk denotes the fact it's a breaking change (see [the top of the file](https://github.com/ocaml/ocaml/blob/trunk/Changes#L4))

kayceesrk replied:
Didn't know that. Reverted this change now.

alainfrisch commented on Makefile:
```diff
@@ -333,12 +331,19 @@ reconfigure:
+utils/domainstate.ml: utils/domainstate.ml.c runtime/caml/domain_state.tbl
+	$(CPP) -I runtime/caml $< > $@
+
+utils/domainstate.mli: utils/domainstate.mli.c utils/domainstate.ml \
```
Very minor: I don't think that this build rule should depend on utils/domainstate.ml.

kayceesrk replied:
Fixed.

alainfrisch commented on utils/domainstate.ml.c:
```diff
@@ -0,0 +1,37 @@
+/**************************************************************************/
+
+#define CAML_CONFIG_H_NO_TYPEDEFS
+#include "config.h"
```
This looks quite fragile, and requires extra care in config.h to avoid generating in text.  Why do we need config.h at all?

Couldn't we generate the .ml/.mli files with grep/awk instead?

kayceesrk replied:
You are right. config.h is not needed. I'll remove it and push a commit.

For the code generation, I am OK with either. Perhaps @stedolan has an opinion since he wrote this code.

nojb commented on asmcomp/amd64/proc.ml:
```diff
@@ -325,6 +325,7 @@ let destroyed_at_oper = function
   | Iop (Iintop_imm(Icheckbound _, _)) when Config.spacetime ->
       [| loc_spacetime_node_hole |]
   | Iswitch(_, _) -> [| rax; rdx |]
+  | Itrywith _ -> [| r11 |]
```
Can you say a few words about this change?

gasche replied:
The `trywith` needs an auxiliary register to load things from memory and push them on stack. Previously it would first push `r14` (the location of the previous trap handler) and then use `r14` as a temporary register (clever!) before setting it to the new trap-handler value (the current value of `rsp`), so it appeared to not use any extra register. With the new register usage scheme, `r14` is the domain state pointer and cannot be used in this way, so a new auxiliary register is needed. It looks like `r11` chosen somewhat arbitrarily, possibly because it is already used by `raise` in a sort of similar fashion.

kayceesrk replied:
@gasche has explained it well. `r11` was chosen among the available registers because it is the last general purpose register that is picked by the register allocator. Moreover, `r11` is not used for OCaml argument passing and is destroyed at c calls and plt stubs (allocations in particular). Hence, it is less likely that the register is alive at a `try_with` compared to say `rax`.

nojb replied:
Got it, thanks; I will need to rebase my risc-v backend on top of this patch so the explanation is helpful!

kayceesrk committed:
Remove inclusion of config.h from the files that generate domain state table.

kayceesrk committed:
Remove unnecessary prerequite from make rule for domainstate.mli

kayceesrk committed:
Unbreak Changes file

alainfrisch commented:
@kayceesrk : sorry to bring back the topic -- I know this is not part of the multicore roadmap -- but being able to run multiple OCaml runtime, with independent heaps, in the same process, would already benefit to many use cases.  If we exclude fork (not reliable with threads anyway) and ignore signals, what exactly remains to be done to achieve that after this PR?  If there is no external C bindings (or if they don't hold domain-local state in global variables), is this only about making the list of global roots part of the domain structure?

alainfrisch commented:
I missed the list of globals -- not yet part of the domain structure -- you provided.  I guess that a bit more work would be needed than just global roots (e.g. for the page table).

kayceesrk commented:
> sorry to bring back the topic 

Happy to discuss. Many of this [list of globlas](https://github.com/ocaml/ocaml/pull/8713#issuecomment-500234301) needs to be moved to domain state; globals roots, major heap structures, ephemeron support, etc. But that is a fairly mechanical change. 

Some work may need to be done for channels, which are shared between the threads. But channels already have support for mutual exclusion thanks to systhreads. So there may not be much work required here. But other than that, I don't see any major roadblocks to having thread-local runtimes.

nojb commented:
> > sorry to bring back the topic
> 
> Happy to discuss. Many of this [list of globlas](https://github.com/ocaml/ocaml/pull/8713#issuecomment-500234301) needs to be moved to domain state; globals roots, major heap structures, ephemeron support, etc. But that is a fairly mechanical change.
> 
> Some work may need to be done for channels, which are shared between the threads. But channels already have support for mutual exclusion thanks to systhreads. So there may not be much work required here. But other than that, I don't see any major roadblocks to having thread-local runtimes.

In case someone was willing to do this work, would it be incompatible with the multicore patches?

kayceesrk commented:
> In case someone was willing to do this work, would it be incompatible with the multicore patches?

The design itself shouldn't preclude the two runtimes, multicore and thread-local, from residing in the same codebase. For example, if the global roots were moved to the domain state, in the case of multicore, each domain's global roots field will point to the same structure in memory whereas they'd point to distinct structures in the case of thread-local runtime. 

I haven't thought through other cases such as multicore + thread-local runtime in the same process (each runtime could itself be multi-threaded, sharing the major heap). If someone was indeed working on this, I'd like to work with them closely to avoid incompatibilities.

damiendoligez commented on Changes:
```diff
@@ -37,14 +37,17 @@ Working version
    OCaml callback when called from C.
    (Jacques-Henri Jourdan, review by Stephen Dolan and Gabriel Scherer)
 
+- #8713: Introduce domain state to hold C global variables in the runtime.
```
When written this way, I can never tell whether it means "to hold all C global variables" or "to hold some C global variables". Maybe that's because English is not my native language?

alainfrisch replied:
You're not the only one to be confused (mandatory reference : https://en.wikipedia.org/wiki/United_Nations_Security_Council_Resolution_242#Interpretations )

dra27 replied:
How about
```
* #8713: Introduce a state table in the runtime to contain the global variables
  which must be duplicated for each domain in the multicore runtime. For native
  code, this domain state is accessible via the exception pointer register (the
  exception pointer now being part of the new domain state).
```
It's probably better to err on the side of marking this a breaking change (although the globals affected are not normally used outside the runtime, I think?). There are a couple of other reviewers who should be in the list too!

kayceesrk replied:
English isn't my first language either. How about "Introduce domain state to hold all the domain-local C global variables in the runtime"?

kayceesrk replied:
> For native code, this domain state is accessible via the exception pointer register

This is only true on amd64. On amd64, the domain state pointer is held in the register which used to hold the exception pointer. On arm, the domain state pointer is held in the register which used to hold the allocation limit. By stealing allocation limit register, there is just one additional load instruction during allocation; pushtrap, poptrap, raise code remains the same. On power, s390x, arm64, there are enough general purpose registers available and hence we take over a general purpose register for domain state pointer. On i386, which has only a few registers, the domain state pointer is maintained in a global. 

Hence, I'd be happy with:

```
* #8713: Introduce a state table in the runtime to contain the global variables which 
must be duplicated for each domain in the multicore runtime.
```

damiendoligez replied:
I'm not sure about marking it as a breaking change. We only do that for changes that have a chance of breaking reasonably-written programs, otherwise we mark everything, and the mark becomes useless.

kayceesrk replied:
Agreed. I'd not consider this to be a breaking change if `compatibility.h` changes are reverted.

kayceesrk replied:
`compatibility.h` changes have been reverted. There are not breaking changes now.

damiendoligez commented on asmcomp/arm64/proc.ml:
```diff
@@ -66,7 +67,7 @@ let register_class r =
 
 let num_available_registers =
-  [| 23; 32 |] (* first 23 int regs allocatable; all float regs allocatable *)
+  [| 22; 32 |] (* first 23 int regs allocatable; all float regs allocatable *)
```
Please update the comment too.

kayceesrk replied:
Fixed.

kayceesrk committed:
Update comment in arm64

kayceesrk committed:
Update Changes file entry

kayceesrk committed:
Revert changes to compatibility.h

kayceesrk committed:
Fix ENDFUNCTION macros args for callbacks

damiendoligez commented on runtime/amd64nt.asm:
```diff
@@ -136,25 +129,24 @@ ENDIF
-    ; Restore caml_young_ptr, caml_exception_pointer
-        mov     r15, caml_young_ptr
-        mov     r14, caml_exception_pointer
+    ; Restore Caml_state_young_ptr, caml_exception_pointer
```
```suggestion
    ; Restore Caml_state(young_ptr)
```

damiendoligez commented on runtime/amd64nt.asm:
```diff
@@ -318,30 +311,28 @@ IFDEF WITH_SPACETIME
         pop     rax
 ENDIF
     ; Setup alloc ptr and exception ptr
```
```suggestion
     ; Setup alloc ptr
```

damiendoligez commented on runtime/arm.S:
```diff
@@ -413,24 +395,21 @@ caml_raise_exn:
+        mov     domain_state_ptr, r0
+    /* Load exception bucket */
+        mov     r0, r1
     /* Reload trap ptr, alloc ptr and alloc limit */
```
```suggestion
     /* Reload trap ptr and alloc ptr */
```

damiendoligez commented on runtime/caml/domain_state.h:
```diff
@@ -0,0 +1,56 @@
+/* This structure sits in the TLS area and is also accessed efficiently
+ * via native code, which is why the indices are important */
+typedef struct {
+#define DOMAIN_STATE(type, name) CAMLalign(8) type name;
```
On 32-bit architectures, this could be `CAMLalign(4)` (with the appropriate changes to the runtime and code generators). Why did you use 8 for all architectures?

kayceesrk replied:
There are a few double values in the domain state (`stat_minor_words`, `stat_promoted_words`, `stat_major_words`, `extra_heap_resources_minor`) which consume 8 bytes even on a 32-bit architecture. It seemed easiest to align everything to 8 bytes rather than work around for doubles.

damiendoligez replied:
You're right, I didn't think of doubles.

damiendoligez commented on runtime/caml/memory.h:
```diff
@@ -355,10 +354,11 @@ CAMLextern struct caml__roots_block *caml_local_roots;  /* defined in roots.c */
 
 #define CAMLxparam1(x) \
   struct caml__roots_block caml__roots_##x; \
+  caml_domain_state* domain_state_##x = Caml_state; \
```
Why not access the fields directly with `Caml_state->` like you do elsewhere?

damiendoligez commented on runtime/caml/domain_state.tbl:
```diff
@@ -0,0 +1,73 @@
+DOMAIN_STATE(asize_t, minor_heap_wsz)
+DOMAIN_STATE(intnat, in_minor_collection)
+DOMAIN_STATE(double, extra_heap_resources_minor)
+DOMAIN_STATE(struct caml_minor_tables*, minor_tables)
```
Could you explain why you are adding an indirection here instead of just having three fields for the three tables?

kayceesrk replied:
Multicore introduces a few more tables and I found it convenient to bunch them together. But I see that it is better to avoid the indirection. Let me change this to three fields.

kayceesrk replied:
I have removed the indirection and introduced three fields for the tables in https://github.com/ocaml/ocaml/pull/8713/commits/7f179daf286eae57ae1bcf2c3374c1b1b147dd40.

damiendoligez commented on runtime/caml/misc.h:
```diff
@@ -445,6 +483,14 @@ extern void caml_instr_atexit (void);
+ * [caml_backtrace_buffer].  In bytecode, it is the same as a
+ * [code_t], in native code it as a [frame_descr *].  The difference
+ * doesn't matter for code outside [backtrace_{byt,nat}.c],
+ * so it is just exposed as a [backtrace_slot].
```
```suggestion
 * so it is just exposed as a [void *].
```

damiendoligez commented on runtime/gen_domain_state32_inc.awk:
```diff
@@ -0,0 +1,36 @@
+#*                                                                        *
+#**************************************************************************
+
+BEGIN{FS=",";count=0};
```
You could do this and simplify your script (the field name is simply $2) while implementing a more permissive syntax:
```suggestion
BEGIN{FS="[,)] *";count=0};
```

kayceesrk replied:
Done. Thanks for the suggestion.

damiendoligez commented on runtime/gen_domain_state64_inc.awk:
```diff
@@ -0,0 +1,33 @@
+#*                                                                        *
+#**************************************************************************
+
+BEGIN{FS=",";count=0};
```
Same here.

kayceesrk replied:
Done.

damiendoligez commented on runtime/meta.c:
```diff
@@ -284,13 +284,7 @@ value caml_static_release_bytecode(value prog, value len)
-value * caml_extern_sp;
-value * caml_trapsp;
-int caml_callback_depth;
+int volatile caml_something_to_do;
```
Why do you need this declaration?

kayceesrk replied:
Removed.

damiendoligez commented on runtime/signals_nat.c:
```diff
@@ -234,15 +234,16 @@ DECLARE_SIGNAL_HANDLER(segv_handler)
        handler, we jump to the asm function [caml_stack_overflow]
        (from $ARCH.S). */
 #ifdef CONTEXT_PC
+    CONTEXT_C_ARG_1 = (context_reg) Caml_state;
```
Shouldn't you change the declaration of `caml_stack_overflow` at line 210 of this file?

kayceesrk replied:
Fixed.

damiendoligez reviewed (commented):
I have carefully reviewed the code and I found only a few nits and some naive questions.

damiendoligez commented:
Note: this patch will conflict with #8805 because they both change the assembler sequence used for allocation. cc @jhjourdan @stedolan @kayceesrk

It will also conflict with #8809 but that should be trivial to fix.

stedolan commented:
> Note: this patch will conflict with #8805 because they both change the assembler sequence used for allocation. cc @jhjourdan @stedolan @kayceesrk

Feel free to review/merge in whatever order's convenient. I'll work on resolving the conflict in whichever PR is merged last. (I don't think the amount of work will change significantly depending on the merge order)

dra27 commented on runtime/caml/domain_state.h:
```diff
@@ -0,0 +1,56 @@
+
+CAMLextern caml_domain_state* Caml_state;
+
+#include "compatibility.h"
```
Note that this doesn't work if `compatibility.h` is actively being used (i.e. it's being included in user code). I'm not sure what the impact is (as in, what code out there relies on `compatibility.h` in order to access the five things which have changed), but this comment is just to track it, and note that it's on my list to propose a fix.

(I don't think it has to block merging)

kayceesrk replied:
I wonder if the easiest thing to do here is to remove the `#include "compatibility.h"` and instead just redefine the undefed macros. `domain_state.h` includes `mlvalues.h` which conditionally includes `compatibility.h` if `CAML_NAME_SPACE` is not defined. Hence, the change will be to delete `#include "compatibility.h"` and replace that with

```c
#ifndef CAML_NAME_SPACE
#define young_start (Caml_state->young_start)                                       
#define young_end (Caml_state->young_end)                                           
#define young_ptr (Caml_state->young_ptr)                                           
#define young_limit (Caml_state->young_limit) 
#endif 
```
We should add comments in `compatibility.h` and `domain_state.h` referring to the other.

kayceesrk replied:
I pushed a fix for this in https://github.com/ocaml/ocaml/pull/8713/commits/239a667d8b6ca559d7b5fa86a88806adc3397c9f. This should leave the behavior of programs that depend on `compatibility.h` unchanged.

kayceesrk committed:
Simplify awk scripts, fix comments and signatures.

kayceesrk committed:
Make depend. Fix reference to ref table.

kayceesrk committed:
Remove a level of indirection for accessing minor tables from Caml_state.

kayceesrk commented:
@damiendoligez I have addressed all of your comments in the recent commits.

damiendoligez reviewed (approved):
Good to go, once you've resolved the discussion about `compatibility.h`.

kayceesrk committed:
Fixes for conflicting definitions in compatibility.h

kayceesrk committed:
Use domain state field names that do not conflict with compatibility.h

The domain state structure (caml_domain_state) uses certaing field names
(young_start, young_limit, etc.) that conflict with the macro
definitions in compatibility.h that use the same names. These macro
definitions are made only when CAML_NAME_SPACE is not defined. To avoid
conflict, when CAML_NAME_SPACE is not defined, we prefix the field names
in caml_domain_state with an underscore.

Update .depend files.

kayceesrk commented:
I have implemented the change suggested by @dra27 to fix the incompatibilities with `compatibility.h` in https://github.com/ocaml/ocaml/pull/8713/commits/ad96dec89c16287b0367cdbc2bb1b83f1d804d32.

dra27 committed:
Propagate Caml_state_field

dra27 committed:
Update Changes

dra27 commented:
Mainly for the record, as @kayceesrk and I have been discussing the last bits offline.

I have a silly test program which provides a primitive for retrieving `caml_young_limit` as a `nativeint` which can be built both with and without `CAML_NAME_SPACE` which I've used to test the compatibility of this. I will try to convert this to an ocamltest shortly.

Compatibility definitions for `caml_young_start`, etc. are now automatically loaded *in user code only* if the appropriate header (`memory.h` or `minor_gc.h`) is `#include`d. Note that this only works outside the compiler source tree (i.e. the compiler codebase itself cannot use these compatibility aliases, but all user code can - we could in the future insert deprecation pragmas, I guess).

`young_start` and so forth works now when `CAML_NAME_SPACE` is not defined because all the fields in Caml_state then begin with an underscore (this is fine, as the `struct` is otherwise identical). So when `CAML_NAME_SPACE` is undefined, `compatibility.h` is able to `#define young_start` to `Caml_state->_young_start` and avoid a recursive definition problem. Fortunately, this ugliness is restricted to the headers only, and a new macro `Caml_state_field` means that in the headers we now just write `Caml_state_field(young_start)`.

dra27 committed:
Compatibility test

dra27 reviewed (approved):
It's making one final pass through [precheck](https://ci.inria.fr/ocaml/job/precheck/293/). Assuming it passes, I see no reason for @stedolan not to push the appropriate merge button...

kayceesrk commented:
The newly added test breaks under MacOS. @dra27 is looking at it.

dra27 committed:
Fix compatibility test recipe

dra27 commented:
It's just completing [precheck#294](https://ci.inria.fr/ocaml/job/precheck/294/), but the macOS tests are already blue, so hopefully it's correct this time!

kayceesrk commented:
Looks like precheck#294 has succeeded. Any thoughts on this @stedolan before pressing the merge button?

stedolan commented:
Looks great! Very happy to see this finally getting in.

gasche commented:
While looking at #8729, I wondered why the global variables in `intern.c` (describing the marshaller state) are not encapsulated in the domain structure as well. Is this intentional? I don't see any of the `intern_*` static variables listed in @kayceesrk's very helpful [table of unchanged globals](https://github.com/ocaml/ocaml/pull/8713#issuecomment-500234301).

kayceesrk commented:
@gasche This looks like an oversight. Thanks for bringing this up. Marshaling has various problems in multicore (https://github.com/ocaml-multicore/ocaml-multicore/issues/75). Since the design isn't final yet, in multicore, we had simply made the globals in `intern.c` thread-local (https://github.com/ocaml-multicore/ocaml-multicore/blob/master/byterun/intern.c#L38-L43). I shall add the global variables in `intern.c` to the table.
